model/pipeline/src/main/proto/beam_runner_api.proto

```diff
@@ -588,6 +588,10 @@ message StandardCoders {
     // of the element
     // Components: The element coder and the window coder, in that order
     WINDOWED_VALUE = 8 [(beam_urn) = "beam:coder:windowed_value:v1"];
+
+    // Encodes an iterable of elements, some of which may be stored in state.
+    // Components: Coder for a single element.
+    STATE_BACKED_ITERABLE = 9 [(beam_urn) = "beam:coder:state_backed_iterable:v1"];
```

Comment:
Instead of this being a state backed iterable, I was hoping that this would be a state backed length prefix coder around any type allowing anything to be truncated. This could replace the length prefix coder or be a new coder which is like an extension of the length prefix coder. After the last chunk of data, the state key is encoded:
Example taken from https://docs.google.com/document/d/1IGduUqmhWDi_69l9nG8kw73HZ5WI5wOps9Tshl5wpQA/edit#heading=h.j8zpiclux4id
Regular length prefix coder (parameterized to 17 bytes):
Length 16 | Data [0, 16) | Length 16 | Data [16, 32) | Length 5 | Data[32, 37)

State backed length prefix coder (parameterized to 16 bytes), we use parameterized length + 1 to signal the state key:
Length 16 | Data [0, 16) | Length 16 | Data [16, 32) | Length 17 | State Key
We can either use a fixed width state key or any length prefix scheme we want such as varint.

This would allow us to drop support for values >2gb over the Data API and allow SDKs to always assume that values are never encoded over multiple data messages simplifying how decoding is done within Runners and SDKs over the Data API.

Reply:
While such a coder may be useful, I don't think that's what we want in this particular case as we want to explicitly let the iterable be lazily read (and written, if it's indeed written rather than just a pointer, say to some shuffle position, stored). The standard iterable coder (as with all other coders) must consume its input to advance the stream to the end of its encoded bytes, which makes it impossible to achieve this kind of composition.

Reply:
I do believe the composition is possible. The state backed length prefix coder would read all the data from the Data API up until the end including the optional state token if it exists. If it hits the end and there is no state token or the component coder doesn't support lazy decoding then it can use the component coder to decode the data and return that result. If there is a state token and the component coder supports a lazy view then it calls a new method on the component coder which creates a lazy view over a reiterable bytestring and its up to the component coder implementation to decode/seek over the data as needed.

I have a partial implementation of the reiterable bytestring in StateFetchingIterators.java, just need to create the lazy decoding iterable view on top of it.

Reply:
Yes, such composition is possible if one extends coder protocol to have the notion of lazy decoding and encoding for component coders, with the all supporting infrastructure, in every SDK. However, I don't think we have the bandwidth to tackle that complexity now, and 95% of the time this is useful is the case of consuming a values iterable from a GBK from the runner. This will allow us to at least have parity with the non-portability framework, and we can look at going beyond that (if the underlying runners support such large records once we pass them across the FnAPI boundary) later. 

I'm happy to annotate this as experimental if that would help.

Reply:
I'll bring this discussion to the mailing list.

Reply:
Yes, I thought about the Data API case to not require chunked values and found out a bunch of subtle complexity there related to making messages still fit within 2gbs since we may make a bad coder choice such as iterable<length_prefix<blob>> instead of length_prefix<iterable<blob>>.

Is there a significant difference in the implementation complexity of having the support for reading state within the length prefix coder while having support for lazy decoding within the iterable portion? (haven't been able to look through the code to understand the nuance)

Yes, marking this experimental to move the implementation forward will make sense but will defer to your judgement and hopefully the ML thread is fruitful.

Reply:
Yes, I think chopping up the byte stream at arbitrary points has many more subtle complexities than splitting things up at element boundaries (for both the data stream and iterable elements), as well as being a non-trivial change to the public API of coders. 

However, thanks for bringing to the list. I'll continue the discussion there; let's see what comes of that.

Reply:
As per the discussion on the list, more general support for large values is desirable but not near as straightforward. I did go ahead and mark this coder as experimental. PTAL.